diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
index 0f6a492..acfff41 100644
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -74,6 +74,8 @@ else
 CHANNEL=$(GUI)
 endif
 TERMINAL=no
+# Support Migemo
+MIGEMO=no
 
 
 # Link against the shared version of libstdc++ by default.  Set
@@ -838,6 +846,10 @@ ifeq (yes, $(STATIC_WINPTHREAD))
 LIB += -Wl,-Bstatic -lwinpthread -Wl,-Bdynamic
 endif
 
+ifeq (yes, $(MIGEMO))
+DEFINES += -DDYNAMIC_MIGEMO
+endif
+
 all: $(TARGET) vimrun.exe xxd/xxd.exe install.exe uninstal.exe GvimExt/gvimext.dll
 
 vimrun.exe: vimrun.c
