diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
index b292720..2f3ec7b 100644
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -73,6 +73,8 @@ CHANNEL=yes
 else
 CHANNEL=$(GUI)
 endif
+# Support Migemo
+MIGEMO=no
 
 
 # Link against the shared version of libstdc++ by default.  Set
@@ -828,6 +830,10 @@ ifeq (yes, $(STATIC_WINPTHREAD))
 LIB += -Wl,-Bstatic -lwinpthread -Wl,-Bdynamic
 endif
 
+ifeq (yes, $(MIGEMO))
+DEFINES += -DDYNAMIC_MIGEMO
+endif
+
 all: $(TARGET) vimrun.exe xxd/xxd.exe install.exe uninstal.exe GvimExt/gvimext.dll
 
 vimrun.exe: vimrun.c
