diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
index 87de4e9..822b014 100644
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -80,6 +80,8 @@ TERMINAL=yes
 else
 TERMINAL=no
 endif
+# Support Migemo
+MIGEMO=no
 
 ifndef CTAGS
 # this assumes ctags is Exuberant ctags
@@ -868,6 +876,10 @@ ifeq (yes, $(STATIC_WINPTHREAD))
 LIB += -Wl,-Bstatic -lwinpthread -Wl,-Bdynamic
 endif
 
+ifeq (yes, $(MIGEMO))
+DEFINES += -DDYNAMIC_MIGEMO
+endif
+
 all: $(TARGET) vimrun.exe xxd/xxd.exe install.exe uninstal.exe GvimExt/gvimext.dll
 
 vimrun.exe: vimrun.c
