diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
index 86bbc3e..acde95f 100644
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -72,6 +72,8 @@ CHANNEL=yes
 else
 CHANNEL=$(GUI)
 endif
+# Support Migemo
+MIGEMO=no
 
 
 # Link against the shared version of libstdc++ by default.  Set
@@ -797,6 +799,10 @@ LIB += -lstdc++
 endif
 endif
 
+ifeq (yes, $(MIGEMO))
+DEFINES += -DDYNAMIC_MIGEMO
+endif
+
 all: $(TARGET) vimrun.exe xxd/xxd.exe install.exe uninstal.exe GvimExt/gvimext.dll
 
 vimrun.exe: vimrun.c
